### DISCUSSION TITLE
Calendar trigger / event

### DISCUSSION THREAD

**Discussion by stefanuytterhoeven** [custom-components/pyscript#358] [Q&A]
In the current version of HA (2022.5+) there's a calendar event AND a calendar trigger.
If you create an automation in HA, you can e.g. when a calender event is found and e.g. 10 minutes before that event (or after that event), you can specify an action.
Since I'm more fond of pyscript, I'm looking for a way to do this in pyscript.
Any suggestion?
One of my use cases:
Situation:
 my wife works one week, from early in the morning till the afternoon, the week after that, she starts around noon till the evening
 calendar (google) with working days added. So monday from 5 'till 13, tuesday from 5 'till 13, .... , the week after that, monday from 13 'till 21 h, tuesday from 13 'till 21h, .....
my wife is leaves about an hour before for her work. 
I want, an hour before her shift starts, lower the temperature at the house. Half an hour after she finishes her shift, raise the temperature again.
So, I want to use an offset. Google calendar supports an offset in the message itself, but that look ugly. And, an automation that could be made in HA itself, is aparently able to use offsets too.
But, as I stated, I'm not a fan of HA automations, and would rather like to do this in pyscript.....

Any ideas someone?

**Comment by gravyflex** [comment]
For example, I have this template sensor and I am trying to do this in pyscript.
```
---

# Daycare Today binary_sensor, Is today a Daycare day for Gabby?
#_________________________________ Daycare ----------------------------------- 

- trigger:
    - platform: calendar
      entity_id: calendar.daycare
      event: start
      offset: -08:40:00
    - platform: calendar
      entity_id: calendar.daycare
      event: end
  binary_sensor:
    - name: Daycare Today
      state: >
        {% if 'Daycare' in trigger.calendar_event.summary %}
          {% if trigger.event == "start" %} 
            on
          {% else %}
            off
          {% endif %}
        {% endif %}
```

**Comment by phsdv** [comment]
I think we are missing support for calendar in pyscript. Maybe someone else can correct me if I am wrong. You can trigger on the events from calendar though, like in below example:
```
@state_trigger("calendar.work_travel")
def calendar_context_test(context=None, **kwargs):
    entity_id = kwargs['var_name']
    log.warning(f"******Entity id: {entity_id}, kwargs: {kwargs}, Context: {context}")
```
resulting in this:
```
******Entity id: calendar.work_travel, kwargs: {'trigger_type': 'state', 'var_name': 'calendar.work_travel', 'value': 'off', 'old_value': 'on'}, Context: <homeassistant.core.Context object at 0x7fdf83b87d30>
******Entity id: calendar.work_travel, kwargs: {'trigger_type': 'state', 'var_name': 'calendar.work_travel', 'value': 'on', 'old_value': 'off'}, Context: <homeassistant.core.Context object at 0x7fdf83387fb0>
```
However I don't know how to access further information from 'context' so it is kind of useless so far.
I also tried @event_trigger, but I never got a {'trigger_type': 'event' back.

**Comment by austinche** [comment]
The `calendar.get_events` service can be used to retrieve a list of events, e.g. all events for the next 24 hours `events = calendar.main.get_events(return_response=True, blocking=True, duration='24:00')`. I use a cron `@time_trigger` to trigger on upcoming events.

**Reply by phsdv** [reply]
I am using calendar.get_events but since 2024.7 I get this error:
``` 
line 39: events = calendar.woonkamer_thermostaat.get_events(return_response=True, blocking=True, duration='00:01') ^ AttributeError: 'StateVal' object has no attribute 'get_events'
```

there was a breaking change in 2024.7:
```
Calendar service list_events was deprecated in favor of get_events in Home Assistant 2023.12 and has now been removed.

Use the new get_events service to retrieve events from calendar entities instead.
```
But I am already using get_events. Any one an idea what is going on? Is pyscript using list_events behind get_events?
Or did core home assistant remove get_events instead of list_events?

**Reply by austinche** [reply]
It still works for me in 2024.7. Where is your calendar coming from? Could whatever integration you're using not have been updated?

**Reply by phsdv** [reply]
I am using the local calendar

**Reply by phsdv** [reply]
update: I upgrade to 2024.7.3 and no more get_events errors.
